### DISCUSSION TITLE
Graph signal processing for recommendation

### DISCUSSION THREAD

**Discussion by vicbeldo98** [pyg-team/pytorch_geometric#4451] [Q&A] [answered]
Hello!

I am a student trying to learn all the ways that recommendation can be made through GNNs. For the moment, I have experimented with PyG and its made examples for recommendation. 

I have seen another approach that uses the similarity of the items and the signals made by the users on those items to make a prediction of the rating of an item based on the signal. [Example](https://gnn.seas.upenn.edu/labs/lab-3/)

My aim is to implement this approach in PyG. The problem is that I do not know how to handle the graph signals inside the convolutions (or if there is a way). 

I have seen there is another library called [Pytorch Geomtric Temporal](https://pytorch-geometric-temporal.readthedocs.io/en/latest/notes/introduction.html) (an extension of this one), but i am unsure if that is what i am looking for, as it seems to focus more on a temporal prediction.

Is there a way to achieve this behaviour on PyG, or PyGT is the solution? I would appreciate some clarification.
Thank you in advanced.

**Comment by rusty1s** [comment, marked as answer]
I am not that familiar with the LocalGNN approach mentioned in the blog, but I see no reason why we cannot replace that layer with any GNN layer from PyG. This indeed looks like a task better handled by PyG rather than PyG Temporal. If you let me know the issues you have in implementing the graph filter, I am happy to help out further.

**Reply by vicbeldo98** [reply]
Okay, I was a bit confused of what to use because of the graph signals mentioned, but I think pytorch geometric temporal is another different topic.

That would be great!
I will post any doubt I have in this issue if that is okay with you.

Thank you very much for all your help!

**Reply by rusty1s** [reply]
Super. Just let me know :)

**Reply by vicbeldo98** [reply]
okay, so I have been looking at the original code in order to understand how to do it. I am going to write a summary just to put things clear and structured. 

LocalGNN, among other parameters, defines two principal inputs:
- **Input x**: set of graph signals of size batchSize x input features x N nodes. Each graph signal represents one user's preferences. If I have four movies (labeled as 0,1,2,3) and I have rated them with [0,3,1,0], we get from there two graph signals: [0,3,0,0] and [0,0,1,0] where our aim is to predict [0,0,1,0] and [0,3,0,0] respectively.
- **Input GSO**: Matrix (1 x N nodes x N nodes) that is a expectation of proximity or similarity between components of the signal. Could be the correlation matrix between movies.
- **Output y**: The prediction for each graph signal (batchSize x dimReadOut[-1] x N nodes)

With those inputs, a neural network with the following structure is built:
![localGNN](https://user-images.githubusercontent.com/49116334/163381200-77fe64cd-ed30-4d27-90ab-88b8e0d9eb2e.png)

The graph filter basically applies the following formula (where k is the specified number of filter taps and h are the learnable parameters):
![image](https://user-images.githubusercontent.com/49116334/163382787-882abba8-395e-4f8e-acb6-cd38923402ad.png)

My problem from here is to adapt the code using PyG. I have seen the [CREATING MESSAGE PASSING NETWORKS](https://pytorch-geometric.readthedocs.io/en/latest/notes/create_gnn.html#creating-message-passing-networks) but i do not know which message is being passed in this approach.
I understand information is being propagated through the similarity matrix of movies, but I don't think we are creating embeddings of the nodes, as the graph is a support structure to propagate info and what we really want to learn are the signals.

Do you have any idea or suggestion of how to adapt this approach in PyG?
Thank you in advanced for your help :)

**Reply by rusty1s** [reply]
It looks like GSO is your edge_index, and h_k is just a scalar. This means that you are applying k message passing iterations, in which messages are simply re-weighted by h_k. Therefore, it is similar to GraphConv where you replace the weight matrix with a single weight.

**Reply by vicbeldo98** [reply]
Okay, I understand 👍 it looks very similar to GraphConv.
Just to clarify, I would have to define the three parameters in GraphConv:
- **edge indices** and **edge weights** that would represent the GSO matrix. This is a matrix of movies so it has 2830 nodes and some edges indicating similarities between movies.
- **node features or X** would be the users graph signals. In the description of GraphConv, it says that its dimensions would be N nodes x Feature dim. My problem then, is how to structure the graph signals, as nodes should correspond to movies (as GSO is movie-based). For example, if thousand people have rated four movies, **X** is a matrix 1000x4. But movies should be rows so i should transpose it like 4x1000. The problem then is that dimensions change between train and test as i could have 4x1000 in train and 4x50 in test.

Let me know if you understood my doubt correctly, I may have complicated the question a bit. And thank you very much for your help and suggestions :) .

**Reply by rusty1s** [reply]
I am not sure how the original approach is doing it, but I think you should represent your features as a num_movies times num_users matrix, e.g. 2830 x 1000. Ultimately, this is a transductive approach so you cannot apply it to new users.

**Reply by vicbeldo98** [reply]
From my understanding you are not doing a recommendation to a particular user, but training to recommend to a generic one. As I see it in this approach they propagate information about similarities between movies to the signals and try to predict a generic signal based on observed behaviours. Let me clarify why I think this is working like that:

You have the matrix GSO (NXN being N the number of movies). This matrix captures the adyacent structure and similarities between movies.

You also have different graph signals that we call x. Hence, x will be a matrix U x N initially (where each row represents a user and each column represents the rating received in that movie for that user ).

As defined previously, GSO * x would be like a one-hop neighboor exchange of information. We are only propagating similarities of movies through the signal and then trying to predict the ones missing.

So, as I see it, given a set of signals, we want to train to predict the ratings missing (not looking into any particular user but rather the movies and its similarities).

In the approach, you can see that they separate training and testing signals, therefore, users are not something critical, rather than its behaviours.
This process is briefly explained [here](https://github.com/alelab-upenn/graph-neural-networks#introduction-).

Let me know if we think the same. I may have gotten something the wrong way.
Thank you very much :)

**Reply by rusty1s** [reply]
Got it. I think we are on the same page. Note that `x` needs to be a `N x U` matrix though (i.e. its transposed).

**Reply by vicbeldo98** [reply]
Good to know we think the same :) 
Then something wrong. I have defined:
- matrix GSO (edge_index 2x 70106 and edge_weights a list of 70106)
- xTrain and yTrain (both 49 x 2830) (users x movies).
- xTest and yTest (both 9 x 2839) 

Looking at the examples made in Pytorch Geometric, I have defined the following architecture:
```
class Encoder(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.conv1 = GraphConv(-1, 64)
        self.conv2 = GraphConv(-1, 64)

    def forward(self, x, edge_index, edge_weights):
        x = self.conv1(x, edge_index, edge_weights).relu()
        x = self.conv2(x, edge_index, edge_weights).relu()
        return x


class Decoder(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.lin1 = Linear(64, 1)

    def forward(self, z_dict):
        z = self.lin1(z_dict).relu()
        return z.view(-1)


class Model(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.encoder = Encoder()
        self.decoder = Decoder()

    def forward(self, x_dict, edge_index_dict, edge_label_index):
        z_dict = self.encoder(x_dict, edge_index_dict, edge_label_index)
        return self.decoder(z_dict)
```

But when I do the following, an error arises:

```
with torch.no_grad():
    model.encoder(xTrain, edge_index, edge_weights)
```
_Index Error: index out of range in self:_
![image](https://user-images.githubusercontent.com/49116334/164412683-e4bcecf9-2b06-44c6-bf31-7f969736a816.png)

 For the contrary, If I transpose xTrain, I get that _TypeError: unsupported operand type(s) for *: 'Tensor' and 'builtin_function_or_method':_
![image](https://user-images.githubusercontent.com/49116334/164412998-d66791ce-35dd-47ae-97ec-6b1adcb9c852.png)

1) can you detect something wrong? probably the model is not correct but it seems to be a problem of the data format (not sure).

2) on the other hand, i dont fully understand why we need N x U, as users in train are different for test, so dimensions will be different?

Thank you in advanced for all your help :)

**Reply by rusty1s** [reply]
1. Can you print `data` and also `edge_index.max()`? How do you transpose the tensor? It looks like you are not calling the function via `x.t()`.
2. I think your `edge_index` need to change accordingly if you apply it to a new node feature matrix.

**Reply by vicbeldo98** [reply]
Okay, I was using x.transpose() instead of x.t() because I though they were the same, but when I tried I was able to see more errors and continue with the implementation.

Reminder of shapes:
data: torch.Size([2830, 49])
edge_index: torch.Size([2, 70106]) (max gives 70106)
edge_weights: torch.Size([70106])

Now after some more fixes I was able to do train with no conflicts. My problems now are two:
1. Dimensions are not the ones I wish:
    I have two graph convolutions and one linear layer.
    First convolution passes from torch.Size([2830, 49]) to torch.Size([2830, 64]).
    Second convolution passes from torch.Size([2830, 64]) to torch.Size([2830, 64]).
   Finally, linear layer passes from torch.Size([2830, 64]) to torch.Size([2830, 1]), that with 
   z.view(-1) returns torch.Size([2830]).

   Would it be posssible that the first convolution passes from torch.Size([2830, 49]) to  
   torch.Size([2830, 49, 64])? I want to transform each rating into a 64 vector and train 
   that. I have tried transforming the data vector into a tensor torch.Size([2830, 49, 1])
 with the unsqueeze but the convolution fails saying RuntimeError: INDICES element is out of DATA bounds, id=49 axis_dim=49. 

```
class Encoder(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.conv1 = GraphConv((-1, -1), 64)

    def forward(self, x, edge_index, edge_weights):
        x = torch.unsqueeze(x, -1)
        print(x.shape)
        x = self.conv1(x, edge_index, edge_weights).relu()
        print(x.shape)
        return x
```
2. I still don't understand how nor why should edge_index change. In the original problem GSO is fixed, it is a matrix NXN that represents similarities between movies, it does not make sense to me that it changes for testing, just because I have less users.

Thank you very very much for your guidance :)

**Reply by rusty1s** [reply]
The original implementation does not have a weight matrix, but only learns a scalar (the importance of the respective hops), that is why you can apply it to arbitrary feature sizes. You may wanna do the same:
```python
class MyConv(MessagePassing)
    def __init__(self):
        super().__init__(aggr='sum')
        self.weight = torch.nn.Parameter(torch.Tensor(1.))
        self.reset_parameters()

    def reset_parameters(self):
        self.weight.data.fill_(1)

    def forward(self, x, edge_index, edge_weight):
        return x + self.weight * self.propagate(edge_index, x=x, edge_weight=edge_weight)
        
    def message(x_j, edge_weight):
        return edge_weight.view(-1, 1) * x_j
```

I am not sure though why you would want to have an embedding of `torch.Size([2830, 49, 64])`. I think you either want to have an embedding per movie or per user, right? The movie/user rating is already fully embedded via the initial matrix: You would just blow up your feature embedding space that way. What's your training task here?

**Reply by vicbeldo98** [reply]
Yes, maybe it is not the best idea to convert the features, I just thought they did it that way.
The task is to predict the missing rating in the given graph signals. Imagine you have seen 4 movies, generating a graph signal of [4,2,1,3]. We would train with four training samples:
[0,2,1,3] => desired prediction for this graph signal ([4,0,0,0])
[4,0,1,3] => desired prediction for this graph signal ([0,2,0,0])
[4,2,0,3] => desired prediction for this graph signal ([0,0,1,0])
[4,2,1,0] => desired prediction for this graph signal ([0,0,0,3])

So if I have as input a torch.Size([2830, 49]) I want as output a vector of the same shape with the predicted ratings.

**Reply by rusty1s** [reply]
In that case, you would apply a classifier shared for each movie and user, right? For example,
```python
out = model(x, edge_index).flatten()
pred = self.lin(out)
loss = F.mse_loss(pred, data.y.flatten())
```

**Reply by vicbeldo98** [reply]
yes, but I don't know how to handle the different incoming numbers of signals.
Train could be torch.Size([2830, 49]) and test torch.Size([2830, 9]) and I need that the model has the ability to predict both. ¿Which should be the dimensions of the linear layer? 
This is the code I have for the moment (i have transposed the matrix because 2830 is common for all) :

``` python
class MyConv(MessagePassing):
    def __init__(self):
        super().__init__(aggr='add')
        self.weight = torch.nn.Parameter(torch.Tensor(1))
        self.reset_parameters()

    def reset_parameters(self):
        self.weight.data.fill_(1)

    def forward(self, x, edge_index, edge_weight):
        return x + self.weight * self.propagate(edge_index, x=x,
                                                edge_weight=edge_weight)

    def message(self, x_j, edge_weight):
        return edge_weight.view(-1, 1) * x_j


class Encoder(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.conv1 = MyConv()
        self.conv2 = MyConv()

    def forward(self, x, edge_index, edge_weights):
        x = self.conv1(x, edge_index, edge_weights).relu()
        x = self.conv2(x, edge_index, edge_weights)
        return x


class Decoder(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.lin1 = Linear(2830, 2830)

    def forward(self, z):
        z = z.t()
        z = self.lin1(z).relu()
        return z.view(-1, 1)


class Model(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.encoder = Encoder()
        self.decoder = Decoder()

    def forward(self, x_dict, edge_index_dict, edge_label_index):
        z_dict = self.encoder(x_dict, edge_index_dict, edge_label_index)
        out = self.decoder(z_dict)
        return out

def movieMSELoss(yHat, y, idxMovie):
    mse = torch.nn.MSELoss()
    return mse(yHat[idxMovie, :],
               y[idxMovie, :])

model = Model()

with torch.no_grad():
    model.encoder(xTrain, edge_index, edge_weights)

optimizer = torch.optim.Adam(model.parameters(), lr=0.1)
scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(optimizer, mode='min',
                                                       patience=10)

print("data: " + str(xTrain.shape))
print("edge_index: " + str(edge_index.shape))
print("edge_weights: " + str(edge_weights.shape))


def train():
    model.train()
    optimizer.zero_grad()
    pred = model(xTrain, edge_index, edge_weights)
    target = yTrain
    loss = movieMSELoss(pred, target, TARGET_MOVIES)
    loss.backward()
    scheduler.step(loss)
    optimizer.step()
    return float(loss)


@torch.no_grad()
def test():
    model.eval()
    pred = model(xTest, edge_index, edge_weights)
    pred = pred.clamp(min=0, max=5)
    target = yTest
    rmse = movieMSELoss(pred, target, TARGET_MOVIES)
    return float(rmse)


for epoch in range(1, 201):
    loss = train()
    test_rmse = test()
    lr = optimizer.state_dict()['param_groups'][0]['lr']
    if epoch % 10 == 0:
        print(f'Epoch: {epoch:03d}, Loss: {loss:.4f}, Test: {test_rmse:.4f}, LR: {lr:.10f}')
```
(ignore TARGET_MOVIES, you can specify the movies to train).
This works but gives a RMSE of 13. If I take out the linear layer it gives a RMSE of 3.5. 
Still in their approach it gives around 1, so I feel like something is wrong. 
Can you see something weird or bad?

Thank you very much!!!! I feel so close of having a model working, I still need to think calmly some of the things we discussed :) but step by step

**Reply by rusty1s** [reply]
This looks pretty good. Transposing the feature matrix is clever, haven't thought about that :)

I think you should remove the `ReLU` in the decoder though, as otherwise, you will lose a bunch of gradients. Let me know if that improves results!

**Reply by vicbeldo98** [reply]
Thanks!
Okay, as my results were non deterministic (because I had different train test split each execution), I have put a seed now.
Now my experiments RMSE are:
1) Encoder (2 convolutions) + decoder ReLU: 8.5
2) Encoder (2 convolutions) + decoder: 6.1
3) Encoder (3 convolutions) + decoder: 6.1
4) Encoder (3 convolutions): 3.6
5) Encoder (2 convolutions): 3.5
6) Encoder (1 convolution): 8.3 (just trying to see a pattern)

Two convolutions seems to be the best and we are still not so close to 1 RMSE. 
The decoder seems to be useless right now and kind of makes sense.
The thing that I don't understand is that they transform ratings into 64 dim vectors each, decode this result again into one number, and obtain 1 of RMSE.

**Reply by rusty1s** [reply]
Couldn't you do the same by using an MLP as decoder? `Sequential(Linear(2830, 64), ReLU(), ...)`

**Reply by vicbeldo98** [reply]
mmmm would that be equivalent? the 2830 are the movies, what I want to transform into a 64 vectors are each of the components of the graph signal, so that the convolution works with that. Looking at the code they have I have found:

``` python
 """
    LocalGNN: implement the selection GNN architecture where all operations are
        implemented locally, i.e. by means of neighboring exchanges only. More
        specifically, it has graph convolutional layers, but the readout layer,
        instead of being an MLP for the entire graph signal, it is a linear
        combination of the features at each node.
"""
```

This LocalGNN calls to GraphFilter(in_features, out_features, filter_taps, edge_features=1, bias=True)
The definition of graph filter is _graph convolutional layer using LSI graph filters_:

``` python
 def forward(self, x):
        # x is of shape: batchSize x dimInFeatures x numberNodesIn
        B = x.shape[0]
        F = x.shape[1]
        Nin = x.shape[2]
        ...
        # Compute the filter output
        u = LSIGF(self.weight, self.S, x, self.bias)
        # So far, u is of shape batchSize x dimOutFeatures x numberNodes
        ...
```

x would be our graph signals. As we are not using batches, it is just  Users x Movies. The linear shift invariant graph filter (LSIGF) is the one in charge of doing the computation of h*GSO*x, and apparently it is the one that learns the hidden dimensions. It is defined as:
``` python
""
    LSIGF(filter_taps, GSO, input, bias=None) Computes the output of a linear
        shift-invariant graph filter on input and then adds bias.
"""
 # Get the parameter numbers:
    F = h.shape[0]
    E = h.shape[1]
    K = h.shape[2]
    G = h.shape[3]
    assert S.shape[0] == E
    N = S.shape[1]
    assert S.shape[2] == N
    B = x.shape[0]
    assert x.shape[1] == G
    assert x.shape[2] == N
    # Or, in the notation we've been using:
    # h in F x E x K x G
    # S in E x N x N
    # x in B x G x N
    # b in F x N
    # y in B x F x N

    # Now, we have x in B x G x N and S in E x N x N, and we want to come up
    # with matrix multiplication that yields z = x * S with shape
    # B x E x K x G x N.
    # For this, we first add the corresponding dimensions
    x = x.reshape([B, 1, G, N])
    S = S.reshape([1, E, N, N])
    z = x.reshape([B, 1, 1, G, N]).repeat(1, E, 1, 1, 1) # This is for k = 0
    # We need to repeat along the E dimension, because for k=0, S_{e} = I for
    # all e, and therefore, the same signal values have to be used along all
    # edge feature dimensions.
    for k in range(1,K):
        x = torch.matmul(x, S) # B x E x G x N
        xS = x.reshape([B, E, 1, G, N]) # B x E x 1 x G x N
        z = torch.cat((z, xS), dim = 2) # B x E x k x G x N
    # This output z is of size B x E x K x G x N
    # Now we have the x*S_{e}^{k} product, and we need to multiply with the
    # filter taps.
    # We multiply z on the left, and h on the right, the output is to be
    # B x N x F (the multiplication is not along the N dimension), so we reshape
    # z to be B x N x E x K x G and reshape it to B x N x EKG (remember we
    # always reshape the last dimensions), and then make h be E x K x G x F and
    # reshape it to EKG x F, and then multiply
    y = torch.matmul(z.permute(0, 4, 1, 2, 3).reshape([B, N, E*K*G]),
                     h.reshape([F, E*K*G]).permute(1, 0)).permute(0, 2, 1)
    # And permute againt to bring it from B x N x F to B x F x N.
    # Finally, add the bias
    if b is not None:
        y = y + b
    return y
```

Now that I see it again it is not learning an embed representation of each rating, but not really sure what it is doing. Seems like a sequence of concatenations, not super sure. I might have to give it a thought.

**Reply by rusty1s** [reply]
Can you clarify the naming of shapes for me? In particular for `E` and `G`? It looks like there exist multiple adjacency matrices `S`  (`E` many) per single node feature matrix.

**Reply by vicbeldo98** [reply]
Okay, sorry for the late response, didn't have time to look until today :(
What I understood is the following:
- Their **x** is not U x N, they reshape **x** such that **x** is now U x 1 x N (understanding it as batch size x dimFeatures x N).
- Also, W is reshaped to E x N x N (where E is like the edge features, which in our case will be always one).

So in every Graph Filter, you pass G (dim_in), F (dim_out), K (filter taps), E (edge_features) and bias.
From these, you construct a tensor h of shape F x E x K x G and define it as torch parameter.
Then pass to the LSIGF four parameters:
- h (F x E x K x G) = (output_features x edge_features x filter_taps x input_features)
- GSO (E x N x N) = (edge_features x number_nodes x number_nodes)
- x (B x G x N) = (batch_size x input_features x number_nodes)
- bias (F x N) = (output_features x number_nodes)

Then after lots and lots of reshaping, it does the normal convolution operation:
``` python
for k in range(1,K):
        x = torch.matmul(x, S) # B x E x G x N
        xS = x.reshape([B, E, 1, G, N]) # B x E x 1 x G x N
        z = torch.cat((z, xS), dim = 2) # B x E x k x G x N
```
Finally, it multiplies the final result z by the parameters and adds bias.
Are you understanding why is it done like this? I kind of understand the code, but not the intentions through every line. What are we learning?

Thank you in advanced for all your help

**Reply by rusty1s** [reply]
Sorry for the late reply. I think I understand the problem now better. It looks like you can achieve the same in PyG via
```python
x = ... [num_nodes, input_features]
edge_index = ... [2, num_edges]

xs = []
for k in range(1, K):
   x = self.propagate(edge_index, x=x)
   xs.append(x)
x = torch.cat(xs, dim=-1)
```
such that the results of intermediate convolution operators are maintained.

**Reply by vicbeldo98** [reply]
Okay, I am going to think this for a bit, thank you very much for all your help :)

**Reply by vicbeldo98** [reply]
Okay, after some time, I reached one of the creators of the library (Fernando Gamma). He clarified some things for me:

```
The idea of the code is to mimic PyTorch as much as possible.

1) In PyTorch, you call nn.Conv2D, you specify input features, output features, and kernel size, and you call your forward on the image (a tensor of shape batch_size x n_features x n_rows x n_cols).
1) In this library, you call GraphFilter, you specify input features, output features, and the number of filter taps (conceptually equivalent to kernel size), and you call your forward on the image (a tensor of shape batch_size x n_features x n_nodes).

2) In PyTorch, you build your architecture calling nn.Conv2D, then you follow with nn.ReLU, then you call nn.Conv2D again, and so on.
2) In this library, you do the same, except you replace nn.Conv2D by GraphFilter.

3) In PyTorch, nn.Conv2D handles the logic of the learnable parameters, but the actual computation is done (in C) by the function nn.functional.conv2d.
3) In this library, GraphFilter handles the logic of the learnable parameters, and the actual computation is done by the function LSIGF (not in C, though)

So, logically speaking, it works exactly the same as convolutions.

The main difference comes as follows:

When you work with images, you know the placement of the pixels: each pixel (except in the border) has one pixel on each side. And this is the same structure for all the images in your data set.
When you work with graph signals, the placement of the nodes is "arbitrary" and specified by a graph. So, unlike nn.Conv2D, which already assumes the structure of the data, you need to specify the data yourself. To do that, you need to define a edge_features x n_nodes x n_nodes tensor that describes the graph (for example, the adjacency matrix will be a 1 x n_nodes x n_nodes matrix; same with the Laplacian). You specify this to your GraphFilter operation by using GraphFilter.addGSO. At first, this structure you specify is the same for all the graph signals in your data set.

In the case of the MovieLens dataset:
- Your nodes are the items. So you take your training set, comprised of users, where each user has rated some of the items. For each pair of items, you take all users that have rated both of those items, and you use that subset to compute the Pearson correlation coefficient between the users. This will amount to your edge weight. You do this with all the pairs, obtaining edge weights between all of your nodes. Then, you apply a k-nearest neighbor to sparsify it. THIS IS YOUR GRAPH. Each node is an item, the edge weights are the Pearson correlation. This graph is fixed and won't change.
- Then, your training set is comprised of "graph signals" which are N-dimensional vector (where N is the number of nodes). Each graph signal corresponds to a user. The user may have rated some of the items, but not others. So in the items the user rated, you put a number between 1 and 5 in the corresponding node, and 0 in all other nodes.
- Then you train.
- And then the test set is build on a similar fashion. The graph has never changed, it's always the same (build from the training set), even for your test set.
```
So it is literal convolution on a graph.

So one last questions:
You have x that are the train signals and S that is the matrix of similarities. Is x * S the same as self.propagate(edge_index, x=x)? x * S gives a new graph signal that is a linear combination of the values of x at neighbours. Just asking to be sure, as self.propagate seems like a  way of sharing embeddings and sum up them, and not sure if they are equivalent.

Thank you for all your responses! You have helped me a lot and I am very grateful for it :) Almost there!

**Reply by rusty1s** [reply]
Yes, this is equivalent. For `aggr="sum"`, `self.propagate(edge_index, x=x)` will perform `A @ X`, where `A` is the sparse + binary adjacency matrix given in `edge_index`, and `X` is the node feature matrix of shape `[num_nodes, num_features]`.

**Reply by vicbeldo98** [reply]
I did a small test and not getting the same results for some reason. Do you see something strange?
``` python
import numpy as np
from torch_geometric.nn import MessagePassing
import torch


x = torch.tensor(np.array([[5,3,2],[0,1,2],[0,5,1],[1,0,4]]).transpose())
edge_index = torch.tensor([[0,2,1,2],[2,0,2,1]])
edge_weights = torch.tensor([0.8, 0.8, 0.2, 0.2])

class MyConv(MessagePassing):
    def __init__(self):
        super().__init__(aggr='sum')

    def forward(self, x, edge_index):
        x = self.propagate(edge_index, x=x)
        return x

    def message(self, x_j):
        return x_j

conv = MyConv()
with torch.no_grad():
    new_x = conv(x, edge_index)
    print(new_x)

S = np.array([[0,0,0.8],[0,0,0.2],[0.8,0.2,0]])
print(np.matmul(S,x))
```

**Reply by rusty1s** [reply]
The results match once you incorporate `edge_weight` into the `MessagePassing` formulation as well:
```python
import numpy as np
from torch_geometric.nn import MessagePassing
import torch

x = torch.tensor(
    np.array([[5, 3, 2], [0, 1, 2], [0, 5, 1], [1, 0, 4]]).transpose())
edge_index = torch.tensor([[0, 2, 1, 2], [2, 0, 2, 1]])
edge_weight = torch.tensor([0.8, 0.8, 0.2, 0.2])


class MyConv(MessagePassing):
    def __init__(self):
        super().__init__(aggr='sum')

    def forward(self, x, edge_index, edge_weight):
        x = self.propagate(edge_index, x=x, edge_weight=edge_weight)
        return x

    def message(self, x_j, edge_weight):
        return x_j * edge_weight.view(-1, 1)


conv = MyConv()
with torch.no_grad():
    new_x = conv(x, edge_index, edge_weight)
    print(new_x)

S = np.array([[0, 0, 0.8], [0, 0, 0.2], [0.8, 0.2, 0]])
print(np.matmul(S, x))
```

**Reply by vicbeldo98** [reply]
Okay, of course, my mind!
I finally have it working!

Here is what the model looks like (if you want to take a look):

``` python

class MyConv(MessagePassing):
    def __init__(self, in_features=1, out_features=64, K=5):
        super().__init__(aggr='add')
        self.K = K
        self.in_features = in_features
        self.out_features = out_features
        self.weight = torch.nn.Parameter(torch.Tensor(self.out_features, self.K, self.in_features))
        #TODO: ADD BIASself.bias = nn.parameter.Parameter(torch.Tensor(F, 1))
        self.reset_parameters()

    def reset_parameters(self):
        stdv = 1. / math.sqrt(self.in_features * self.K)
        self.weight.data.uniform_(-stdv, stdv)
        '''if self.bias is not None:
            self.bias.data.uniform_(-stdv, stdv)'''

    def forward(self, x, edge_index, edge_weight):
        # conv dimensions == B * F_in * K * N
        conv = x.permute(1, 0).reshape([-1, self.in_features, 1, N_movies])

        for k in range(1, self.K):
            x = self.propagate(edge_index, x=x, edge_weight=edge_weight)
            x_aux = x.permute(1, 0).reshape([-1, self.in_features, 1, N_movies])
            conv = torch.cat((conv, x_aux), dim=2)

        # Actually multiply by the parameters
        # Reshape conv must be KG x F ===> order to B x N_movies x K x in_features and  reshape to B x N x (K*in_features)
        reshaped_conv = conv.permute(0, 3, 2, 1).reshape([-1, N_movies, self.K*self.in_features])

        # h convert KG x F
        h = self.weight.reshape([self.out_features, self.K*self.in_features]).permute(1, 0)

        y = torch.matmul(reshaped_conv, h)
        return y

    def message(self, x_j, edge_weight):
        return x_j * edge_weight.view(-1, 1)


class Encoder(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.conv1 = MyConv()

    def forward(self, x, edge_index, edge_weights):
        x = self.conv1(x, edge_index, edge_weights).relu()
        return x


class Decoder(torch.nn.Module):
    def __init__(self, out_features=64, dim_readout=1):
        super().__init__()
        self.lin1 = Linear(out_features, dim_readout)

    def forward(self, z):
        z = self.lin1(z)
        z = z.reshape(-1, N_movies).permute(1,0)
        return z


class Model(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.encoder = Encoder()
        self.decoder = Decoder()

    def forward(self, x_dict, edge_index_dict, edge_label_index):
        y = self.encoder(x_dict, edge_index_dict, edge_label_index)
        z_dict = self.decoder(y)
        return z_dict
```
Basically, I am just doing the same as they are but using the propagate of PyG.
This is like doing a convolution on the signals.
My lowest RMSE arrives now to 0.97 (even though of coure this is just learning for one movie).
But this was what I wanted to mimic.

Thank you very much for all your help! Not only with code but also for helping me understand was was done!

**Reply by rusty1s** [reply]
Amazing :) Glad that this is finally working :)
